{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set DMZ working directory, all paths will be relative (downstream) to this directory\n",
    "\n",
    "import os\n",
    "\n",
    "working_directory = \"../../data/Su_2020_FAIR\"  #TODO: parameterize?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.listdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Read in Patient Metadata\n",
    "\n",
    "import pandas as pd\n",
    "patient_metadata = pd.read_excel('../../data/Su_2020_original/IDs_Individuals-vs-Samples_incl-all-CONTROLS.xlsx')\n",
    "patient_metadata = patient_metadata.rename(columns={\"Study Subject ID\": \"Study_Subject_ID\",  \"Sample ID\": \"Sample_ID\", \"BD Time-Point\": \"BD_Time_point\"})\n",
    "patient_metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set Sample IDs to strings (some control samples are integers)\n",
    "patient_metadata['Sample_ID'] = patient_metadata['Sample_ID'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from isatools.model import *\n",
    "ontologies = {\n",
    "    \"CHEBI\": OntologySource(\n",
    "        name = \"CHEBI - Chemical Entities of Biological Interest\", \n",
    "        file = \"http://purl.obolibrary.org/obo/chebi.owl\",\n",
    "        description = \"A structured classification of molecular entities of biological interest focusing on 'small' chemical compounds.\"), \n",
    "    \"CHMO\": OntologySource(\n",
    "        name = \"CHMO - Chemical Methods Ontology\", \n",
    "        file = \"http://purl.obolibrary.org/obo/chmo.owl\",\n",
    "        description = \"CHMO, the chemical methods ontology, describes methods used to collect data in chemical experiments, such as mass spectrometry and electron microscopy prepare and separate material for further analysis, such as sample ionisation, chromatography, and electrophoresis synthesise materials, such as epitaxy and continuous vapour deposition It also describes the instruments used in these experiments, such as mass spectrometers and chromatography columns. It is intended to be complementary to the Ontology for Biomedical Investigations (OBI).\"), \n",
    "    \"CRO\": OntologySource(\n",
    "        # The Contributor Role Ontology (CRO) is an extension of the CASRAI Contributor Roles Taxonomy (CRediT) and replaces the former Contribution Ontology.\n",
    "        name = \"CRO - Contributor Role Ontology\",\n",
    "        file = \"http://purl.obolibrary.org/obo/cro.owl\",\n",
    "        description = \"A classification of the diverse roles performed in the work leading to a published research output in the sciences. Its purpose to provide transparency in contributions to scholarly published work, to enable improved systems of attribution, credit, and accountability.\"),\n",
    "    \"EDAM\": OntologySource(\n",
    "        name = \"EDAM - EMBRACE Data and Methods\",\n",
    "        file = \"http://edamontology.org/EDAM.owl\",\n",
    "        description = \"Bioinformatics operations, data types, formats, identifiers and topics\"),  \n",
    "    \"EFO\": OntologySource(\n",
    "        name = \"EFO - Experimental Factor Ontology\", \n",
    "        file = \"http://www.ebi.ac.uk/efo/efo.owl\",\n",
    "        description = \"The Experimental Factor Ontology (EFO) provides a systematic description of many experimental variables available in EBI databases, and for external projects such as the NHGRI GWAS catalogue. It combines parts of several biological ontologies, such as anatomy, disease and chemical compounds. The scope of EFO is to support the annotation, analysis and visualization of data handled by many groups at the EBI and as the core ontology for OpenTargets.org\"), \n",
    "    \"MI\": OntologySource(\n",
    "        name = \"MI - Molecular Interactions Controlled Vocabulary\", \n",
    "        file = \"http://purl.obolibrary.org/obo/mi.owl\",\n",
    "        description = \"A structured controlled vocabulary for the annotation of experiments concerned with protein-protein interactions.\"),\n",
    "    \"MS\": OntologySource(\n",
    "        name = \"MS - Mass spectrometry ontology\",\n",
    "        file = \"http://purl.obolibrary.org/obo/ms.owl\",\n",
    "        description = \"A structured controlled vocabulary for the annotation of experiments concerned with proteomics mass spectrometry.\"),\n",
    "    \"MSIO\": OntologySource(\n",
    "        name = \"MS - Mass spectrometry ontology\",\n",
    "        file = \"http://purl.obolibrary.org/obo/msio.owl\",\n",
    "        description = \"MSIO aims to provide a single point of entry to support semantic markup of experiments making use of NMR and MS techniques to identify, measure and quantify small molecules known as metabolites. MSIO covers metabolite profiling, targeted or undertargeted, tracer based applications. MSIO reuses a number of resources such as CHEBI, DUO, NMRCV, OBI, and STATO.\"),\n",
    "    \"NCBITAXON\": OntologySource(\n",
    "        name = \"NCBI organismal classification\", \n",
    "        file = \"http://purl.obolibrary.org/obo/ncbitaxon.owl\",\n",
    "        description = \"An ontology representation of the NCBI organismal taxonomy\"),\n",
    "    \"NCIT\": OntologySource(\n",
    "        name = \"NCI Thesaurus OBO Edition\", \n",
    "        file = \"http://purl.obolibrary.org/obo/ncit.owl\",\n",
    "        description = \"The NCIt OBO Edition project aims to increase integration of the NCIt with OBO Library ontologies. NCIt is a reference terminology that includes broad coverage of the cancer domain, including cancer related diseases, findings and abnormalities. NCIt OBO Edition releases should be considered experimental.\"),\n",
    "    \"OBI\": OntologySource(\n",
    "        name = \"OBI - Ontology for Biomedical Investigations\", \n",
    "        file = \"http://purl.obolibrary.org/obo/obi.owl\",\n",
    "        description = \"An integrated ontology for the description of life-science and clinical investigations\"),\n",
    "    \"OMIABIS\": OntologySource(\n",
    "        name = \"Ontologized MIABIS\", \n",
    "        file = \"http://purl.obolibrary.org/obo/omiabis.owl\",\n",
    "        description = \"An ontological version of MIABIS (Minimum Information About BIobank data Sharing)\"),\n",
    "    \"PRIDE\": OntologySource(\n",
    "        name = \"PRIDE Controlled Vocabulary\",\n",
    "        file = \"http://purl.obolibrary.org/obo/pride_cv.obo\",\n",
    "        description = \"The PRIDE PRoteomics IDEntifications (PRIDE) database is a centralized, standards compliant, public data repository for proteomics data, including protein and peptide identifications, post-translational modifications and supporting spectral evidence.\"),\n",
    "   \"UBERON\": OntologySource(\n",
    "        name = \"Uber-anatomy ontology\",\n",
    "        file = \"http://purl.obolibrary.org/obo/uberon.owl\",\n",
    "        description = \"Uberon is an integrated cross-species anatomy ontology representing a variety of entities classified according to traditional anatomical criteria such as structure, function and developmental lineage. The ontology includes comprehensive relationships to taxon-specific anatomical ontologies, allowing integration of functional, phenotype and expression data.\")\n",
    "    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "investigation = Investigation(\n",
    "    filename = \"i_investigation.txt\", \n",
    "    identifier = \"\", \n",
    "    title = \"Trusted World of Corona (TWOC)\",\n",
    "    description = \"The goal of this trusted guide to the world of COVID-19 is to help clinicians, the scientific community, policy makers and politicians and the public at large to get near real time accurate, expert-annotated and specific information in a modern, user friendly and easily accessible format. The benefits will include better use of treatments, faster development of vaccines and a clearer view on factors that may negatively affect the outcomes of a COVID-19 infection and other future virus outbreaks. Website: https://www.health-holland.com/project/2020/trusted-world-of-corona\",\n",
    "    submission_date = \"\",\n",
    "    public_release_date = \"\",\n",
    "    ontology_source_references = [o for o in ontologies.values()],\n",
    "     publications = [\n",
    "        Publication(doi=\"https://doi.org/10.31219/osf.io/9mz27\", \n",
    "                    title='Mild as well as severe disease caused by COVID-19 might be part of the same problem: Machine-assisted analysis of congruent clinical observations and the underlying molecular mechanisms in order to rationalise drug repurposing',\n",
    "                    status=OntologyAnnotation(\n",
    "                                term=\"preprint\",\n",
    "                                term_source= ontologies[\"EFO\"],\n",
    "                                term_accession=\"http://www.ebi.ac.uk/efo/EFO_0010558\"),\n",
    "                    author_list=\"Barend Mons, Peter-Bram 't Hoen, Dirkjan Kuijpers, Thomas Hankemeier, Gianpiero Pescarmona\")],\n",
    "    \n",
    "    contacts = [\n",
    "        Person(\n",
    "            last_name = \"Mons\", \n",
    "            first_name = \"Barend\",\n",
    "            affiliation = \"\",\n",
    "            email = \"info@twoc.eu\",\n",
    "            address = \"http://www.twoc.eu/\",\n",
    "            roles = [\n",
    "                OntologyAnnotation(\n",
    "                    term = \"project management role\",\n",
    "                    term_source = ontologies[\"CRO\"], \n",
    "                    term_accession =\"http://purl.obolibrary.org/obo/CRO_0000065\")])],\n",
    "    studies = None,\n",
    "    comments = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort_study = Study(\n",
    "    filename = \"s_study.txt\", \n",
    "    identifier = \"https://doi.org/10.1016/j.cell.2020.10.037\", \n",
    "    title = \"Multi-Omics Resolves a Sharp Disease-State Shift between Mild and Moderate COVID-19\",\n",
    "    description = \"Multi-omics and clinical of patients with mild, moderate and severe corona infection\", \n",
    "    submission_date = \"\", \n",
    "    public_release_date = \"\",\n",
    "    contacts = [\n",
    "        Person(\n",
    "            last_name = \"Su\", \n",
    "            first_name = \"Yanpeng\",\n",
    "            #mid_initials = \"\",\n",
    "            affiliation = \"Institute for Systems Biology, Seattle, WA 98109, USA\",\n",
    "            roles = [\n",
    "                OntologyAnnotation(\n",
    "                    term = \"author role\",\n",
    "                    term_source = ontologies[\"CRO\"], \n",
    "                    term_accession =\"http://purl.obolibrary.org/obo/CRO_0000001\")]),\n",
    "        Person(\n",
    "            last_name = \"Heath\", \n",
    "            first_name = \"James\",\n",
    "            mid_initials = \"R.\",\n",
    "            affiliation = \"Institute for Systems Biology, Seattle, WA 98109, USA\",\n",
    "            roles = [\n",
    "                OntologyAnnotation(\n",
    "                    term = \"author role\",\n",
    "                    term_source = ontologies[\"CRO\"], \n",
    "                    term_accession =\"http://purl.obolibrary.org/obo/CRO_0000001\")])],\n",
    "    design_descriptors = [\n",
    "        OntologyAnnotation(\n",
    "                term = \"Multi-omics study\",\n",
    "                term_source = ontologies[\"PRIDE\"],\n",
    "                term_accession = \"http://purl.obolibrary.org/obo/PRIDE_0000461\"),\n",
    "        OntologyAnnotation(\n",
    "                term = \"population based study design\",\n",
    "                term_source = ontologies[\"OMIABIS\"],\n",
    "                term_accession = \"http://purl.obolibrary.org/obo/OMIABIS_0001022\")],\n",
    "    publications = [\n",
    "        Publication(doi=\"https://doi.org/10.1016/j.cell.2020.10.037\", pubmed_id= '33171100',\n",
    "                    title='Multi-Omics Resolves a Sharp Disease-State Shift between Mild and Moderate COVID-19',\n",
    "                    status=OntologyAnnotation(term=\"indexed in PubMed\"),\n",
    "                    author_list=\"Su Y, Chen D, Yuan D, Lausted C, Choi J, Dai CL, Voillet V, Duvvuri VR, Scherler K, Troisch P, Baloni P, Qin G, Smith B, Kornilov SA, Rostomily C, Xu A, Li J, Dong S, Rothchild A, Zhou J, Murray K, Edmark R, Hong S, Heath JE, Earls J, Zhang R, Xie J, Li S, Roper R, Jones L, Zhou Y, Rowen L, Liu R, Mackay S, O'Mahony DS, Dale CR, Wallick JA, Algren HA, Zager MA; ISB-Swedish COVID19 Biobanking Unit, Wei W, Price ND, Huang S, Subramanian N, Wang K, Magis AT, Hadlock JJ, Hood L, Aderem A, Bluestone JA, Lanier LL, Greenberg PD, Gottardo R, Davis MM, Goldman JD, Heath JR\")],\n",
    "    factors = None, \n",
    "    protocols = None,\n",
    "    assays = None,\n",
    "    sources = None,\n",
    "    samples = None,\n",
    "    process_sequence = None,\n",
    "    other_material = None,\n",
    "    characteristic_categories = None,\n",
    "    comments = None,\n",
    "    units = None)\n",
    "investigation.studies.append(cohort_study)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define protocol parameters \n",
    "protocol_params = {\n",
    "    \"Post Extraction\": ProtocolParameter(\n",
    "        parameter_name = \"Post Extraction\"\n",
    "        ),\n",
    "\n",
    "    \"Derivatization\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"Derivatization\",\n",
    "            term_source = ontologies[\"MSIO\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/MSIO_0000111\")\n",
    "        ),\n",
    "\n",
    "    \"Chromatography Instrument\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"Chromatography Instrument\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0000485\")\n",
    "        ),\n",
    "\n",
    "    \"Column model\": ProtocolParameter(\n",
    "        parameter_name = \"Column model\"\n",
    "        ),\n",
    "\n",
    "    \"Column type\": ProtocolParameter(\n",
    "        parameter_name = \"Column type\"\n",
    "        ),\n",
    "\n",
    "    \"Scan polarity\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"scan polarity\",\n",
    "            term_source = ontologies[\"MS\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/MS_1000465\")\n",
    "        ),\n",
    "        \n",
    "    \"Scan m/z range\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"Scan m/z range\")\n",
    "        ),\n",
    "            #term_source = ontologies[\"\"],\n",
    "            #term_accession = \"\")),\n",
    "        \n",
    "    \"Instrument\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"Instrument\",\n",
    "            term_source = ontologies[\"MS\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/MS_1000463\")\n",
    "        ),\n",
    "\n",
    "    \"Ion source\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"Ion source\",\n",
    "            term_source = ontologies[\"CHMO\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/CHMO_0000960\")\n",
    "        ),\n",
    "\n",
    "    \"Mass analyzer\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"Mass analyzer\",\n",
    "            term_source = ontologies[\"MS\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/MS_1000451\")\n",
    "        ),\n",
    "\n",
    "    \"method reference\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"method reference\",\n",
    "            term_source = ontologies[\"MI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/MI_0357\")\n",
    "        ),\n",
    "\n",
    "    \"Panel amount\": ProtocolParameter(\n",
    "        parameter_name = \"Panel amount\"\n",
    "        ),\n",
    "\n",
    "    \"Panel size\": ProtocolParameter(\n",
    "        parameter_name = \"Panel size\"\n",
    "        ),\n",
    "        \n",
    "    \"PEA Instrument\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"Proximity extension assay\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0003113\")\n",
    "        ),\n",
    "\n",
    "    \"PCR instrument\": ProtocolParameter(\n",
    "        parameter_name = OntologyAnnotation(\n",
    "            term = \"real-time PCR machine\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0001110\")\n",
    "        )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define Assays\n",
    "\n",
    "assays = {\n",
    "    \"metabolomics\": Assay(\n",
    "        filename = \"a_assay_metabolomics.txt\",\n",
    "        measurement_type = OntologyAnnotation(\n",
    "            term = \"targeted metabolite profiling\",\n",
    "            term_source = ontologies[\"MSIO\"], \n",
    "            term_accession = \"http://purl.obolibrary.org/obo/MSIO_0000100\"),\n",
    "    \n",
    "        technology_type = OntologyAnnotation(\n",
    "            term = \"liquid chromatography-mass spectrometry\",\n",
    "            term_source = ontologies[\"CHMO\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/CHMO_0000524\")\n",
    "        ),\n",
    "\n",
    "    \"proteomics\": Assay(\n",
    "        filename = \"a_assay_proteomics.txt\",\n",
    "        measurement_type = OntologyAnnotation(\n",
    "            term = \"plasma proteomics assay\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0003229\"),\n",
    "\n",
    "        technology_type = OntologyAnnotation(\n",
    "            term = \"Proximity extension assay\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0003113\")\n",
    "        ),\n",
    "\n",
    "    \"transcriptomics\": Assay(\n",
    "        filename = \"a_assay_transcriptomics.txt\",\n",
    "        measurement_type = OntologyAnnotation(\n",
    "            term = \"single-cell RNA sequencing assay\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0002631\"),\n",
    "\n",
    "        technology_type = OntologyAnnotation(\n",
    "            term = \"single-cell RNA sequencing assay\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0002631\")\n",
    "        )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define protocols\n",
    "\n",
    "protocols = {\n",
    "    \n",
    "    \"sample_collection\": Protocol( \n",
    "        name = \"sample_collection_protocol\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"sample collection\", # alternative term, original is material sampling process\n",
    "            term_source = ontologies[\"OBI\"],  \n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0000744\"),\n",
    "\t#TODO: Blood draw; first, second draw\n",
    "\t#TODO: Plasma and PBMC isolation\n",
    "       # metabolomics protocols\n",
    "        description = \"WOS for time of blood draw were determined by manual expert review. WOS for Figure S1A were automatically generated from data extracted from the electronic health record for hospitalized patients, and plotted for 6-hour time intervals based on end-interval grade. Automated results were compared against manual expert review for 15% of study subjects. The follng data were collected from the subject’s electronic health record (EHR): complete blood count (CBC) with differential, comprehensive metabolic panel, APTT, D-dimer, fibrinogen, prothrombin time, thrombin time ,and troponin I. Lab data were extracted from the nearest time point to each blood draw, if available within a window ± two days. First blood draw (n = 76), second blood draw (n = 54). Blood draws were classified as WOS = 3-4 (n = 83) and WOS = 5-7 (n = 47). We used an unpaired Wilcoxon-test to determine the statistical difference between WOS = 3-4 and WOS = 5-7, and P values were FDR adjusted. Spearman correlation coefficient was calculated using R package ‘corrplot v0.84’ to observe the associations between EHR labs and WOS disease severity, and the correlation significance was reported as FDR adjusted P values.\"),\n",
    "    \n",
    "    \"Extraction\": Protocol(\n",
    "        name = \"Extraction\", \n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"extraction\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0302884\"), #also in: MSIO, etc.\n",
    "        parameters = [protocol_params[\"Post Extraction\"],\n",
    "                      protocol_params[\"Derivatization\"]],\n",
    "        description = \"Plasma and PBMC isolation were conducted with standard protocols from Bloodworks Northwest (Seattle, WA). Patient blood was collected in BD Vacutainer (EDTA) tubes (Becton, Dickinson and Company, Franklin Lakes, NJ). Plasma fractions were collected after centrifuged at 800 x g at 4°C for 10 min, aliquoted, and stored until use at −80°C. The rest of the blood was diluted with PBS (pH7.2) to 2X of the original volume and layered over 15 mL Ficoll (GE Healthcare, Waukesha, WI) in SepMate-50 tubes (StemCell, Vancouver, BC). After centrifuged at 800 x g for 15 min at room temperature, the PBMC layer (did not include granulocytes (such as neutrophils)) was poured into a 50 mL conical tube. The cells were washed twice with autoMACS Rinsing Solution (Miltenyi Biotec, Auburn, CA) and centrifuge at 250 x g for 10 min, at RT. PBMC pellets were gently resuspended in 5 mL Rinsing Solution and a 5 microLiter aliquot was diluted 1:10 v/v for cell counting. Cells in 18 microLiter of diluted samples were first mixed with 2 microLiter of Acridine Orange / Propidium Iodide Stain (Logos Biosystems, Annandale, VA), 10 microLiter was then loaded to a PhotonSlide (Logos Biosystems) and counted in a LUNA FL Dual Fluorescence cell counter (Logos Biosystems). Cryopreservation freeze media CryoStor CS-10 (Biolife Solutions, Bothell, WA) was slowly added to make a concentration of 2.5 million PBMC/ml. Cells were aliquoted in 2.0 mL Cryotube vials (ThermoFisher, Waltham, MA) and frozen in CoolCell LX Cell Freezing Container (Corning, Corning, NY) at −80°C for at least 2 hours before stored in liquid nitrogen until use.\"),\n",
    "    \n",
    "    \"Labelling samples\": Protocol(\n",
    "        name = \"Labelling samples\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = 'Labelling',\n",
    "            term_source = ontologies[\"CHMO\"],\n",
    "            term_accession = 'http://purl.obolibrary.org/obo/CHMO_0001675')\n",
    "        ),\n",
    "    \n",
    "    \"Chromatography\": Protocol(\n",
    "        name = \"Chromatography\", \n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"chromatography\",\n",
    "            term_source = ontologies[\"CHMO\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/CHMO_0001000\"), \n",
    "        parameters = [protocol_params[\"Chromatography Instrument\"],\n",
    "                      protocol_params[\"Column model\"],\n",
    "                      protocol_params[\"Column type\"]],\n",
    "        description = \"Metabolon (Morrisville, NC, USA) conducted the metabolomics assays for all participant plasma samples used in this study. Data were generated with the Global Metabolomics platform via ultra-high-performance liquid chromatography/tandem accurate mass spectrometry. 100 microLiter of plasma was aliquoted and transported on dry ice to Metabolon Inc. for analysis. Sample handling and quality control were performed by Metabolon in their CLIA-certified laboratory. Mass spectrometry was performed using Metabolon’s ultra-high-performance liquid chromatography/tandem mass spectrometry (UHPLC/MS/MS) Global Platform, which consisting of four independent UPLC–MS/MS instruments, each with a Waters ACQUITY ultra-performance liquid chromatography (UPLC) and Thermo Scientific Q-Exactive high resolution/accurate mass spectrometer interfaced with a heated electrospray ionization (HESI-II) source and Orbitrap mass analyzer at 35,000 mass resolution.\"),\n",
    "    \n",
    "    \"Mass spectrometry\": Protocol(\n",
    "        name = \"Mass spectrometry\", \n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"mass spectrometry\",\n",
    "            term_source = ontologies[\"CHMO\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/CHMO_0000470\"), \n",
    "        parameters = [protocol_params[\"Scan polarity\"],\n",
    "                      protocol_params[\"Scan m/z range\"],\n",
    "                      protocol_params[\"Instrument\"],\n",
    "                      protocol_params[\"Ion source\"],\n",
    "                      protocol_params[\"Mass analyzer\"],\n",
    "                      protocol_params[\"method reference\"]]\n",
    "        ),\n",
    "    \n",
    "    \"Data transformation\": Protocol(\n",
    "        name = \"Data transformation\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"data transformation\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0200000\"),\n",
    "        description = \"Raw metabolomics data were median scaled within each batch such that the median value for each metabolite was one\"\n",
    "        ),\n",
    "\n",
    "    \"metabolite identification\": Protocol(\n",
    "        name = 'metabolite identification',\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = 'metabolite identification',\n",
    "            term_source = ontologies[\"MI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/MI_2131\")   \n",
    "        ),\n",
    "    \n",
    "    \"Proximity extension assay\": Protocol(\n",
    "        name = \"Proximity extension assay\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"Proximity extension assay\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0003113\")\n",
    "        ),\n",
    "        \n",
    "    \"real-time PCR\": Protocol(\n",
    "        name = \"real-time PCR\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"real-time PCR instrument\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0001110\"),\n",
    "        ),\n",
    "    \n",
    "    \"feature annotation\": Protocol(\n",
    "        name = \"feature annotation\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"Annotation\",\n",
    "            term_source = ontologies[\"EDAM\"],\n",
    "            term_accession = \"http://edamontology.org/operation_0226\")\n",
    "        ),    \n",
    "\n",
    "    \"PBMC preparation\": Protocol(\n",
    "        name = \"PBMC preparation\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"staining\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0302887\"),\n",
    "        description = \"Chromium Single Cell Kits (10x Genomics) were utilized to analyze the transcriptomic, surface protein levels and, TCR sequences simultaneously from the same cell. Experiments were performed according to the manufacturer’s instructions. Briefly, cryopreserved PBMCs were thawed and 1X red blood cell lysis solution (BioLegend) was used to lyse any remaining red blood cells in the PBMC samples. Cells were stained with a panel of TotalSeq-C human antibodies that includes hashtag multiplexing antibodies (BioLegend) detailed in Table S1.8 according to the manufacturer’s protocol.\"\n",
    "        ),\n",
    "\n",
    "    \"Single cell sequencing\": Protocol(\n",
    "        name = \"Single cell sequencing\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"DNA sequencing assay\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0000626\"),\n",
    "        description =  \"Stained cells were then loaded onto a Chromium Next GEM chip G. Cells were lysed for reverse transcription and complementary DNA (cDNA) amplification in the Chromium Controller (10X Genomics). The polyadenylated transcripts were reverse-transcribed inside each gel bead-in-emulsion afterward. Full-length cDNA along with cell barcode identifiers were PCR-amplified and sequencing libraries were prepared and normalized. The constructed library was sequenced on the Novaseq platform (Illumina).\"\n",
    "        ),\n",
    "\n",
    "     \"Single cell RNA-seq data qc\": Protocol(\n",
    "         name = \"Single cell RNA-seq data qc\",\n",
    "         protocol_type = OntologyAnnotation(\n",
    "            term = \"quality control testing design\",\n",
    "            term_source = ontologies[\"OBI\"],\n",
    "            term_accession = \"http://purl.obolibrary.org/obo/OBI_0001336\"),\n",
    "         description =  \"Droplet-based sequencing data were aligned and quantified using the Cell Ranger Single-Cell Software Suite (version 3.0.0, 10x Genomics) against the GRCh38 human reference genome. Cells from each demultiplexed sample were first filtered for cells that expressed a minimum of 200 genes, then they were filtered based on three metrics: 1) the total number of unique molecular identifiers (UMI) counts per cell (library size) must be less than 10000; 2) the number of detected genes per cell must be less than 2500; and 3) the proportion of mitochondrial gene counts (UMIs from mitochondrial genes / total UMIs) must be less than 10%. Doublets were either simultaneously identified in sample demultiplexing or identified using scrublet (Wolock et al., 2019) and were removed prior to the aforementioned filtering. After QC metric filtering, a total of 559,583 cells were retained for downstream analysis. Scanpy (Wolf et al., 2018) was used to normalize cells via CPM normalization (UMI total count of each cell was set to 106) and log1p transformation (natural log of CPM plus one).\"\n",
    "        ),\n",
    "\n",
    "     \"Pseudo-bulk aggregation\": Protocol(\n",
    "        name = \"Pseudo-bulk aggregation\",\n",
    "        protocol_type = OntologyAnnotation(\n",
    "            term = \"pseudo-bulk aggregation of single-cell expression data\",\n",
    "            term_source = ontologies[\"EFO\"],\n",
    "            term_accession =  \"http://www.ebi.ac.uk/efo/EFO_0030053\")\n",
    "        #description = \"TODO\"\n",
    "        )\n",
    "\n",
    "\n",
    "     \n",
    "        \n",
    "    # TODO: Single cell RNA-seq cell type identification?\n",
    "    # TODO: Single cell RNA-seq batch information?\n",
    "    # TODO: Single cell RNA-seq gene regulatory networks?\n",
    "    # TODO: Single cell RNA-seq signature score?\n",
    "    # TODO: Single cell RNA-seq marker selection?\n",
    "       \n",
    "}\n",
    "\n",
    "\n",
    "# append to study protocols\n",
    "for protocol in protocols.values():\n",
    "    cohort_study.protocols.append(protocol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add samples\n",
    "for index, row in patient_metadata.iterrows():\n",
    "    \n",
    "    # create source (=individual)\n",
    "    source_name = row[\"Study_Subject_ID\"]\n",
    "    source = Source(\n",
    "        name = source_name,\n",
    "        characteristics = [\n",
    "            Characteristic(\n",
    "                category = OntologyAnnotation(\n",
    "                    term = \"Organism\",\n",
    "                    term_source = ontologies[\"OBI\"],\n",
    "                    term_accession = \"http://purl.obolibrary.org/obo/OBI_0100026\"),\n",
    "                value = OntologyAnnotation(\n",
    "                    term = \"Homo sapiens\",\n",
    "                    term_source = ontologies[\"NCBITAXON\"],\n",
    "                    term_accession = \"http://purl.obolibrary.org/obo/NCBITaxon_9606\"))])\n",
    "    \n",
    "    if row[\"BD_Time_point\"] == 'healthy donor':\n",
    "        source.characteristics.append(\n",
    "                Characteristic(\n",
    "                    category = OntologyAnnotation(\n",
    "                        term = \"COVID-19 Infection\",\n",
    "                        term_source = ontologies[\"NCIT\"],\n",
    "                        term_accession = \"http://purl.obolibrary.org/obo/NCIT_C171133\"),\n",
    "                    value = OntologyAnnotation(\n",
    "                        term = \"No\",\n",
    "                        term_source = ontologies[\"NCIT\"],\n",
    "                        term_accession = \"http://purl.obolibrary.org/obo/NCIT_C49487\")))\n",
    "        cohort_study.sources.append(source)\n",
    "    \n",
    "    else:\n",
    "        source.characteristics.append(\n",
    "                Characteristic(\n",
    "                    category = OntologyAnnotation(\n",
    "                        term = \"COVID-19 Infection\",\n",
    "                        term_source = ontologies[\"NCIT\"],\n",
    "                        term_accession = \"http://purl.obolibrary.org/obo/NCIT_C171133\"),\n",
    "                    value = OntologyAnnotation(\n",
    "                        term = \"Yes\",\n",
    "                        term_source = ontologies[\"NCIT\"],\n",
    "                        term_accession = \"http://purl.obolibrary.org/obo/NCIT_C49488\")))\n",
    "        cohort_study.sources.append(source)\n",
    "\n",
    "    \n",
    "    \n",
    "    # create sample\n",
    "    sample_name = row[\"Sample_ID\"]\n",
    "    blood_sample = Sample(\n",
    "        name = sample_name, \n",
    "        derives_from = [source])\n",
    "    blood_sample.characteristics.append(\n",
    "        Characteristic(\n",
    "            category = OntologyAnnotation(\n",
    "                term = \"anatomical entity\",\n",
    "                term_source = ontologies[\"UBERON\"],\n",
    "                term_accession = \"http://purl.obolibrary.org/obo/UBERON_0001062\"),\n",
    "            value = OntologyAnnotation(\n",
    "                term = \"blood\",\n",
    "                term_source = ontologies[\"UBERON\"],\n",
    "                term_accession = \"http://purl.obolibrary.org/obo/UBERON_0000178\")))\n",
    "    cohort_study.samples.append(blood_sample)\n",
    "    \n",
    "    # Sample collection_process                \n",
    "    sample_collection_process = Process(\n",
    "        name = \"samplecollection_{0}\".format(row[\"Study_Subject_ID\"]),\n",
    "        executes_protocol = protocols[\"sample_collection\"],\n",
    "        inputs = [source],\n",
    "        outputs = [blood_sample])\n",
    "    cohort_study.process_sequence.append(sample_collection_process)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Add samples to metabolomics assay\n",
    "\n",
    "\n",
    "# Metabolomics subdirectory\n",
    "metabolomics_dir = os.path.join(working_directory, \"metabolomics\")\n",
    "\n",
    "# Datafiles (not all are available)\n",
    "raw_datafile = DataFile(filename= \"blank\", label = \"Raw Spectral Data File\")\n",
    "normalized_datafile = DataFile(filename= \"\", label = \"Normalization Name\")\n",
    "derived_spectral_datafile = DataFile(filename= \"\", label = \"Derived Spectral Data File\")\n",
    "transformation_datafile = DataFile(filename= os.path.join(metabolomics_dir, \"metabolomics_Su_2020_feature-data.csv\"), label = \"Derived Transformation Name\")\n",
    "MAF = DataFile(filename= os.path.join(metabolomics_dir, \"metabolomics_Su_2020_feature-metadata.csv\"), label=\"Metabolite Assignment File\")\n",
    "\n",
    "# Iterate over samples in study\n",
    "for index, row in patient_metadata.iterrows():\n",
    "    for idx, sample in enumerate(cohort_study.samples):\n",
    "        if row[\"Metabol.\"] == 'Y':\n",
    "            if sample.name == row[\"Sample_ID\"]:\n",
    "\n",
    "                # Metabolomics extraction\n",
    "                material_extract = Material(\n",
    "                            name = \"extract_{0}\".format(sample.name),\n",
    "                            type_ = \"Extract Name\"\n",
    "                        )\n",
    "\n",
    "                Post_extraction = ParameterValue(\n",
    "                            category = protocol_params[\"Post Extraction\"], \n",
    "                            value = \"100microLiter of plasma was aliquoted and transported on dry ice\"\n",
    "                        )\n",
    "\n",
    "                Derivatization = ParameterValue(\n",
    "                            category = protocol_params[\"Derivatization\"], \n",
    "                            value = \"Handled by Metabolon Inc.\"\n",
    "                        ) # Metabolon?\n",
    "\n",
    "                extraction_process = Process(\n",
    "                            executes_protocol=protocols[\"Extraction\"], \n",
    "                            parameter_values=[Post_extraction, Derivatization],\n",
    "                            inputs = [sample],\n",
    "                            outputs = [material_extract]\n",
    "                        )\n",
    "\n",
    "                ## Labeling\n",
    "                material_label = Material(\n",
    "                            name =\"labeled_{0}\".format(sample.name),\n",
    "                            type_ =\"Labeled Extract Name\"\n",
    "                        )\n",
    "\n",
    "                labelling_process = Process(\n",
    "                            executes_protocol=protocols[\"Labelling samples\"],\n",
    "                            inputs = [extraction_process.outputs[0]],\n",
    "                            outputs = [material_label]\n",
    "                        )\n",
    "\n",
    "                ## Chromatography\n",
    "                chromatography_instrument = ParameterValue(\n",
    "                            category = protocol_params[\"Chromatography Instrument\"], \n",
    "                            value = \"Waters ACQUITY HPLC\"\n",
    "                        )\n",
    "\n",
    "                column_model = ParameterValue(\n",
    "                            category = protocol_params[\"Column model\"], \n",
    "                            value = \"\"\n",
    "                        ) #TODO unspecified in article???\n",
    "\n",
    "                column_type = ParameterValue(\n",
    "                            category = protocol_params[\"Column type\"], \n",
    "                            value = \"\"\n",
    "                        ) #TODO unspecified in article???\n",
    "\n",
    "                chromatography_process = Process(\n",
    "                            name = \"chromatography_{0}\".format(sample.name),\n",
    "                            executes_protocol = protocols[\"Chromatography\"],\n",
    "                            parameter_values = [chromatography_instrument, column_model, column_type],\n",
    "                            inputs = [labelling_process.outputs[0]],\n",
    "                            outputs = []\n",
    "                        )\n",
    "\n",
    "                ## Mass spectrometry\n",
    "                scan_polarity = ParameterValue(\n",
    "                            category = protocol_params[\"Scan polarity\"], \n",
    "                            value = \"\"\n",
    "                        ) #TODO unspecified in article\n",
    "\n",
    "                scan_range = ParameterValue(\n",
    "                            category = protocol_params[\"Scan m/z range\"], \n",
    "                            value = \"\"\n",
    "                        ) #TODO unspecified in article\n",
    "\n",
    "                ms_instrument = ParameterValue(\n",
    "                            category = protocol_params[\"Instrument\"], \n",
    "                            value = \"Thermo Scientific Q-Exactive\"\n",
    "                        )\n",
    "\n",
    "                ion_source = ParameterValue(\n",
    "                            category = protocol_params[\"Ion source\"], \n",
    "                            value = \"HESI-II\"\n",
    "                        )\n",
    "\n",
    "                mass_analyzer = ParameterValue(\n",
    "                            category = protocol_params[\"Mass analyzer\"], \n",
    "                            value = \"Orbitrap\"\n",
    "                        )\n",
    "\n",
    "                mass_spectrometry_process = Process(\n",
    "                            name = \"mass_spectrometry_{0}\".format(sample.name),\n",
    "                            executes_protocol= protocols[\"Mass spectrometry\"],\n",
    "                            parameter_values = [scan_polarity, scan_range, ms_instrument, ion_source, mass_analyzer],\n",
    "                            inputs = [],\n",
    "                            outputs = [raw_datafile]\n",
    "                        )\n",
    "\n",
    "                ## Data transformation\n",
    "                data_transformation_process = Process(\n",
    "                            name = \"data_transformation_{0}\".format(sample.name),\n",
    "                            executes_protocol = protocols[\"Data transformation\"],\n",
    "                            inputs = [raw_datafile],\n",
    "                            outputs = [normalized_datafile, derived_spectral_datafile]  \n",
    "                        )\n",
    "\n",
    "                ## Metabolite identification\n",
    "                metabolite_identification_process = Process(\n",
    "                            name = \"metabolite_identification_{0}\".format(sample.name),\n",
    "                            executes_protocol =  protocols[\"metabolite identification\"],\n",
    "                            inputs = [normalized_datafile],\n",
    "                            outputs= [transformation_datafile, MAF]                 \n",
    "                        )\n",
    "\n",
    "\n",
    "                # Link processes\n",
    "                plink(extraction_process, labelling_process)\n",
    "                plink(labelling_process, chromatography_process)\n",
    "                plink(chromatography_process, mass_spectrometry_process)\n",
    "                plink(mass_spectrometry_process, data_transformation_process)\n",
    "                plink(data_transformation_process, metabolite_identification_process)\n",
    "\n",
    "\n",
    "                # Add samples, materials and data files to the amines assay\n",
    "                assays[\"metabolomics\"].samples.append(blood_sample)\n",
    "                assays[\"metabolomics\"].other_material.append(material_extract)\n",
    "                assays[\"metabolomics\"].other_material.append(material_label)\n",
    "                assays[\"metabolomics\"].data_files.append(raw_datafile)\n",
    "                assays[\"metabolomics\"].data_files.append(normalized_datafile)\n",
    "                assays[\"metabolomics\"].data_files.append(derived_spectral_datafile)\n",
    "                assays[\"metabolomics\"].data_files.append(transformation_datafile)                                                                                                   \n",
    "                assays[\"metabolomics\"].data_files.append(MAF)\n",
    "\n",
    "\n",
    "                ## Add processes to the amines assay\n",
    "                assays[\"metabolomics\"].process_sequence.append(extraction_process)\n",
    "                assays[\"metabolomics\"].process_sequence.append(labelling_process)\n",
    "                assays[\"metabolomics\"].process_sequence.append(chromatography_process)\n",
    "                assays[\"metabolomics\"].process_sequence.append(mass_spectrometry_process)\n",
    "                assays[\"metabolomics\"].process_sequence.append(data_transformation_process)\n",
    "                assays[\"metabolomics\"].process_sequence.append(metabolite_identification_process)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add samples to proteomics assay\n",
    "\n",
    "\n",
    "\n",
    "# Proteomics subdirectory\n",
    "proteomics_dir = os.path.join(working_directory, \"proteomics\")\n",
    "\n",
    "# Datafiles (not all are available)\n",
    "raw_datafile = DataFile(filename= os.path.join(proteomics_dir, \"proteomics_Su_2020_feature-data.csv\"), label = \"Raw Spectral Data File\")\n",
    "feature_annotation_file = DataFile(filename= os.path.join(proteomics_dir, \"proteomics_Su_2020_feature-metadata.csv\"), label = \"Feature Annotation File\")\n",
    "\n",
    "\n",
    "\n",
    "# Iterate over samples in study\n",
    "for index, row in patient_metadata.iterrows():\n",
    "    for idx, sample in enumerate(cohort_study.samples):\n",
    "        if row[\"Prote.\"] == 'Y':\n",
    "            if sample.name == row[\"Sample_ID\"]:\n",
    "\n",
    "                # proteomics extraction\n",
    "                material_extract = Material(\n",
    "                            name = \"extract_{0}\".format(sample.name),\n",
    "                            type_ = \"Extract Name\"\n",
    "                        )\n",
    "\n",
    "                Post_extraction = ParameterValue(\n",
    "                            category = protocol_params[\"Post Extraction\"], \n",
    "                            value = \"100microLiter of plasma was aliquoted and transported on dry ice\"\n",
    "                        )\n",
    "\n",
    "                Derivatization = ParameterValue(\n",
    "                            category = protocol_params[\"Derivatization\"], \n",
    "                            value = \"\"\n",
    "                        ) #TODO unspecified in articles\n",
    "\n",
    "                extraction_process = Process(\n",
    "                            executes_protocol=protocols[\"Extraction\"], \n",
    "                            parameter_values=[Post_extraction, Derivatization],\n",
    "                            inputs = [sample],\n",
    "                            outputs = [material_extract]\n",
    "                        )\n",
    "\n",
    "                ## Labeling\n",
    "                material_label = Material(\n",
    "                            name =\"labeled_{0}\".format(sample.name),\n",
    "                            type_ =\"Labeled Extract Name\"\n",
    "                        )\n",
    "\n",
    "                labelling_process = Process(\n",
    "                            executes_protocol=protocols[\"Labelling samples\"],\n",
    "                            inputs = [extraction_process.outputs[0]],\n",
    "                            outputs = [material_label]\n",
    "                        )\n",
    "\n",
    "                ## Proximity extension assay\n",
    "                pea_instrument = ParameterValue(\n",
    "                            category = protocol_params[\"PEA Instrument\"], \n",
    "                            value = \"Olink Bioscience\"\n",
    "                        )\n",
    "\n",
    "                Panel_size = ParameterValue(\n",
    "                            category = protocol_params[\"Panel size\"], \n",
    "                            value = \"92\"\n",
    "                        )\n",
    "\n",
    "                Panel_amount = ParameterValue(\n",
    "                            category = protocol_params[\"Panel amount\"], \n",
    "                            value = \"5\"\n",
    "                        )\n",
    "\n",
    "                pea_process = Process(\n",
    "                            name = \"PEA_{0}\".format(sample.name),\n",
    "                            executes_protocol = protocols[\"Proximity extension assay\"],\n",
    "                            parameter_values = [pea_instrument, Panel_size, Panel_amount],\n",
    "                            inputs = [labelling_process.outputs[0]],\n",
    "                            outputs = []\n",
    "                        )\n",
    "\n",
    "                ## real-time PCR\n",
    "                pcr_instrument = ParameterValue(\n",
    "                            category = protocol_params[\"PCR instrument\"], \n",
    "                            value = \"high-throughput microfludic real-time PCR\"\n",
    "                        )\n",
    "\n",
    "                rt_pcr_process = Process(\n",
    "                            name = \"real-time PCR_{0}\".format(sample.name),\n",
    "                            executes_protocol= protocols[\"real-time PCR\"],\n",
    "                            parameter_values = [pcr_instrument],\n",
    "                            inputs = [],\n",
    "                            outputs = [raw_datafile]\n",
    "                        )\n",
    "\n",
    "                ## feature annotation\n",
    "                feature_annotation_process = Process(\n",
    "                            name = \"feature_annotation_{0}\".format(sample.name),\n",
    "                            executes_protocol = protocols[\"feature annotation\"],\n",
    "                            inputs = [raw_datafile],\n",
    "                            outputs = [feature_annotation_file]\n",
    "\n",
    "                )\n",
    "\n",
    "\n",
    "                # Link processes\n",
    "                plink(extraction_process, labelling_process)\n",
    "                plink(labelling_process, pea_process)\n",
    "                plink(pea_process, rt_pcr_process)\n",
    "                plink(rt_pcr_process, feature_annotation_process)\n",
    "\n",
    "\n",
    "                # Add samples, materials and data files to the amines assay\n",
    "                assays[\"proteomics\"].samples.append(blood_sample)\n",
    "                assays[\"proteomics\"].other_material.append(material_extract)\n",
    "                assays[\"proteomics\"].other_material.append(material_label)\n",
    "                assays[\"proteomics\"].data_files.append(raw_datafile)\n",
    "                assays[\"proteomics\"].data_files.append(feature_annotation_file)\n",
    "\n",
    "\n",
    "                ## Add processes to the amines assay\n",
    "                assays[\"proteomics\"].process_sequence.append(extraction_process)\n",
    "                assays[\"proteomics\"].process_sequence.append(labelling_process)\n",
    "                assays[\"proteomics\"].process_sequence.append(pea_process)\n",
    "                assays[\"proteomics\"].process_sequence.append(rt_pcr_process)\n",
    "                assays[\"proteomics\"].process_sequence.append(feature_annotation_process)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add samples to transcriptomics assay\n",
    "\n",
    "\n",
    "\n",
    "# Transcriptomics subdirectory\n",
    "transcriptomics_dir = os.path.join(working_directory, \"transcriptomics\")\n",
    "\n",
    "# Datafiles (not all are available)\n",
    "pseudobulked_file = DataFile(filename= os.path.join(transcriptomics_dir, \" transcriptomics_Su_2020_feature-data.csv\"), label = \"Pseudobulk File\")\n",
    "feature_annotation_file = DataFile(filename= os.path.join(transcriptomics_dir, \"transcriptomics_Su_2020_feature-metadata.csv\"), label = \"Feature Annotation File\")\n",
    "\n",
    "\n",
    "\n",
    "# Iterate over samples in study\n",
    "for index, row in patient_metadata.iterrows():\n",
    "    for idx, sample in enumerate(cohort_study.samples):\n",
    "        if row[\"Transcript.\"] == 'Y':\n",
    "            if sample.name == row[\"Sample_ID\"]:\n",
    "      \n",
    "\n",
    "                # transcriptomics extraction\n",
    "                material_extract = Material(\n",
    "                            name = \"extract_{0}\".format(sample.name),\n",
    "                            type_ = \"Extract Name\"\n",
    "                        )\n",
    "\n",
    "                Post_extraction = ParameterValue(\n",
    "                            category = protocol_params[\"Post Extraction\"], \n",
    "                            value = \" Cells were aliquoted in 2.0 mL Cryotube vials (ThermoFisher, Waltham, MA) and frozen in CoolCell LX Cell Freezing Container (Corning, Corning, NY) at −80°C for at least 2 hours before stored in liquid nitrogen until use.\"\n",
    "                        )\n",
    "\n",
    "                Derivatization = ParameterValue(\n",
    "                            category = protocol_params[\"Derivatization\"], \n",
    "                            value = \"\"\n",
    "                        ) #TODO unspecified in articles\n",
    "\n",
    "\n",
    "                extraction_process = Process(\n",
    "                            executes_protocol=protocols[\"Extraction\"], \n",
    "                            #parameter_values=[Post_extraction, Derivatization],\n",
    "                            inputs = [sample],\n",
    "                            outputs = [material_extract]\n",
    "                        )\n",
    "\n",
    "\n",
    "#                 # Single cell transcriptomics cell preparation\n",
    "#                 material_label = Material(\n",
    "#                             name = \"labeled_{0}\".format(sample.name),\n",
    "#                             type_ = \"Labeled Extract Name\"\n",
    "#                         )\n",
    "\n",
    "#                 cell_preparation_process = Process(\n",
    "#                             executes_protocol=protocols[\"Labelling samples\"], \n",
    "#                             inputs = [material_extract],\n",
    "#                             outputs = [material_label]\n",
    "#                         )\n",
    "\n",
    "\n",
    "                # Sequencing\n",
    "\n",
    "                sc_sequences_raw_file = DataFile(\n",
    "                    filename = \"scs_rawdata_{0}_R1.tsv\".format(material_extract.name), \n",
    "                    label = \"Raw Data File R1\", \n",
    "                    generated_from = [material_extract])\n",
    "\n",
    "                sequencing_process = Process(\n",
    "                            executes_protocol=protocols[\"Single cell sequencing\"], \n",
    "            #                 parameter_values=[feature_metadata_file],\n",
    "                            inputs = [material_extract],\n",
    "                            outputs = [sc_sequences_raw_file]\n",
    "                        )\n",
    "\n",
    "\n",
    "                # Sequencing data QC\n",
    "\n",
    "                sc_sequences_qc_file = DataFile(\n",
    "                    filename = \"scs_qc_{0}.tsv\".format(material_extract.name), \n",
    "                    label = \"Single-cell Sequencing Raw Data File\", \n",
    "                    generated_from = [sc_sequences_raw_file])\n",
    "\n",
    "                sequencing_data_qc_process = Process(\n",
    "                            executes_protocol=protocols[\"Single cell RNA-seq data qc\"], \n",
    "                            inputs = [sc_sequences_raw_file],\n",
    "                            outputs = [sc_sequences_qc_file]\n",
    "\n",
    "                        )\n",
    "\n",
    "                # Pseudo-bulk aggregation\n",
    "\n",
    "                pseudobulk_process = Process(\n",
    "                            executes_protocol=protocols[\"Pseudo-bulk aggregation\"],\n",
    "                            inputs = [sc_sequences_qc_file],\n",
    "                            outputs = [pseudobulked_file]\n",
    "                ) \n",
    "\n",
    "\n",
    "                 ## feature annotation\n",
    "                feature_annotation_process = Process(\n",
    "                            name = \"feature_annotation_{0}\".format(material_extract.name),\n",
    "                            executes_protocol = protocols[\"feature annotation\"],\n",
    "                            inputs = [pseudobulked_file],\n",
    "                            outputs = [feature_annotation_file]\n",
    "                )\n",
    "\n",
    "\n",
    "                # Link processes\n",
    "#                 plink(extraction_process, cell_preparation_process)\n",
    "                plink(extraction_process, sequencing_process)\n",
    "                plink(sequencing_process, sequencing_data_qc_process)\n",
    "                plink(sequencing_data_qc_process, pseudobulk_process)\n",
    "                plink(pseudobulk_process, feature_annotation_process)\n",
    "\n",
    "                # Add samples, materials and data files to the amines assay\n",
    "                assays[\"transcriptomics\"].samples.append(blood_sample)\n",
    "                assays[\"transcriptomics\"].other_material.append(material_extract)\n",
    "#                 assays[\"transcriptomics\"].other_material.append(material_label)\n",
    "                assays[\"transcriptomics\"].data_files.append(sc_sequences_raw_file)\n",
    "                assays[\"transcriptomics\"].data_files.append(sc_sequences_qc_file)\n",
    "                assays[\"transcriptomics\"].data_files.append(pseudobulked_file)\n",
    "                assays[\"transcriptomics\"].data_files.append(feature_annotation_file)\n",
    "\n",
    "\n",
    "                ## Add processes to the  assay\n",
    "                assays[\"transcriptomics\"].process_sequence.append(extraction_process)\n",
    "#                 assays[\"transcriptomics\"].process_sequence.append(cell_preparation_process)\n",
    "                assays[\"transcriptomics\"].process_sequence.append(sequencing_process)\n",
    "                assays[\"transcriptomics\"].process_sequence.append(sequencing_data_qc_process)\n",
    "                assays[\"transcriptomics\"].process_sequence.append(pseudobulk_process)\n",
    "                assays[\"transcriptomics\"].process_sequence.append(feature_annotation_process)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add assays to cohort study\n",
    "for assay in assays.values():\n",
    "        cohort_study.assays.append(assay)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write ISA-Tab files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write to ISA-Tab\n",
    "from isatools import isatab\n",
    "isatab.dump(investigation, working_directory)\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write to ISA-json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "with open(os.path.join(working_directory, \"isa.json\"), \"w\") as out_file:\n",
    "    json.dump(\n",
    "        investigation, \n",
    "        out_file,\n",
    "        cls = ISAJSONEncoder, \n",
    "        sort_keys = True, \n",
    "        indent = 4, \n",
    "        separators = (',', ': '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.6 ('python3isa')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "8dc67bcfc54321ed9774b432913c7e6eb9fda909d4295e73f8e0cfce319139b9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
